crates/rustc_codegen_spirv/src/linker/structurizer.rs

```diff
@@ -228,6 +229,217 @@ fn retarget_loop_children_if_needed(blocks: &mut [Block], cf_info: &ControlFlowI
     }
 }
 
+fn incoming_edges(id: Word, blocks: &[Block]) -> Vec<Word> {
+    let mut incoming_edges = Vec::new();
+    for block in blocks {
+        let out = outgoing_edges(block);
+        if out.contains(&id) {
+            incoming_edges.push(block.label_id().unwrap());
+        }
+    }
+
+    incoming_edges
+}
+
+fn num_incoming_edges(id: Word, blocks: &[Block]) -> usize {
+    incoming_edges(id, blocks).len()
+}
+
+// Turn a block into a conditional branch that either goes to yes or goes to merge.
+fn change_block_to_switch(
+    blocks: &mut [Block],
+    block_id: Word,
+    cases: &[Word],
+    merge: Word,
+    condition: Word,
+) {
+    let cb_idx = find_block_index_from_id(blocks, &block_id);
+    let cb_block = &mut blocks[cb_idx];
+
+    // selection merge
+    let merge_operands = vec![
+        Operand::IdRef(merge),
+        Operand::SelectionControl(SelectionControl::NONE),
+    ];
+    *cb_block.instructions.last_mut().unwrap() =
+        Instruction::new(Op::SelectionMerge, None, None, merge_operands);
+
+    // conditional branch
+    let mut switch_operands = vec![Operand::IdRef(condition), Operand::IdRef(merge)];
+
+    for (i, label_id) in cases.iter().enumerate() {
+        let literal = i as u32 + 1;
+        switch_operands.push(Operand::LiteralInt32(literal));
+        switch_operands.push(Operand::IdRef(*label_id));
+    }
+
+    cb_block
+        .instructions
+        .push(Instruction::new(Op::Switch, None, None, switch_operands));
+}
+
+fn find_or_create_int_constant(
+    opcode: Op,
+    constants: &mut Vec<Instruction>,
+    header: &mut ModuleHeader,
+    type_result_id: Word,
+    value: u32,
+) -> Word {
+    // create
+    let result_id = id(header);
+    let new_constant = Instruction::new(
+        opcode,
+        Some(type_result_id),
+        Some(result_id),
+        vec![Operand::LiteralInt32(value)],
+    );
+    constants.push(new_constant);
+
+    result_id
+}
+
+fn find_or_create_type_constant(
+    opcode: Op,
+    constants: &mut Vec<Instruction>,
+    header: &mut ModuleHeader,
+) -> Word {
+    // find
+    for constant in constants.iter() {
+        if constant.class.opcode == opcode {
+            return constant.result_id.unwrap();
+        }
+    }
+
+    // create
+    let result_id = id(header);
+    let new_constant = Instruction::new(opcode, None, Some(result_id), vec![]);
+    constants.push(new_constant);
+
+    result_id
+}
+
+// detect the intermediate break block by checking whether a block that branches to a merge block has 2 parents.
+fn defer_loop_internals(
+    header: &mut ModuleHeader,
+    blocks: &mut Vec<Block>,
+    cf_info: &ControlFlowInfo,
+    debug_names: &mut Vec<(Word, String)>,
+    constants: &mut Vec<Instruction>,
+) {
+    for loop_info in &cf_info.loops {
```

Comment:
this is pretty far from what the final algorithm should be, but, it's fine for now